Add config for handling obj normalization with missing __dask_tokenize__

hristog opened the pull request:
- [x] Closes #6555
  - [x] Added `tokenize.allow-random` config, accepting `True` or `False`.
- [x] Tests added / passed
  - [x] Added `test_normalize_object` to `dask/tests/test_base.py`.
- [x] Passes `black dask` / `flake8 dask`

hristog commented on dask/tests/test_base.py:
```diff
@@ -8,6 +8,8 @@
 
 from tlz import merge, partial, compose, curry
 
+from unittest import mock
```
Not entirely sure what convention to employ here in terms of whether to `import mock` at this level or within `test_normalize_object` itself (just as it's been done for `import warnings`; the latter has been imported within the test only, because its general utility is not as wide, in this particular context).

Any feedback would be greatly appreciated.

hristog commented on dask/base.py:
```diff
@@ -839,7 +839,42 @@ def normalize_object(o):
+
+        import warnings
+
+        warnings.warn(
```
This implicitly displays a `UserWarning`. Would a different type of warning be more suitable here (e.g., `RuntimeWarning`)?

hristog commented on dask/base.py:
```diff
@@ -839,7 +839,42 @@ def normalize_object(o):
+        return normalized
+
+    if handle_missing_dask_tokenize_method == "raise":
+        raise TypeError(missing_dask_tokenize_method_msg)
```
I've opted for raising a `TypeError` here. Please, let me know if you believe some other type would be more suitable for this circumstance (e.g., `ValueError`).

hristog commented on dask/base.py:
```diff
@@ -839,7 +839,42 @@ def normalize_object(o):
+    if handle_missing_dask_tokenize_method == "raise":
+        raise TypeError(missing_dask_tokenize_method_msg)
+
+    raise ValueError(
```
I've opted to not raise this, if the object is a `callable` or has an associated `__dask_tokenize__` method.

hristog commented:
Need to go through the rest of the codebase a bit more to get an idea how it'd be best to go about documenting the newly-proposed config.

hristog commented:
I'm considering documenting the config as part of [this](https://docs.dask.org/en/latest/custom-collections.html?highlight=normalize_token#implementing-deterministic-hashing) section, because I haven't been able to find a more suitable one.
Any suggestions would be greatly appreciated.

jrbourbeau commented on dask/tests/test_base.py:
```diff
@@ -224,6 +226,88 @@ def test_normalize_base():
         assert normalize_token(i) is i
 
 
+def test_normalize_object():
```
What do you think about something like (again I've not tested the snippet below):

```python
with dask.config.set({"tokenize.allow-random": False}):
    with pytest.raises(RuntimeError, match="..."):
        tokenize(object())
```

jrbourbeau reviewed (commented):
Thanks for the PR @hristog!

We can probably scale the changes back here a bit to just support a boolean flag for whether or not we should raise an error. I suspect a relatively small group of users will utilize this setting and when things are non-deterministic they'll want an error to be raised. We can always expand thing in the future if needed. 

Something like (I've not tested the snippet below)

```python
if callable(o):
    return normalize_function(o)
elif dask.config.get("tokenize.allow-random", True):
    return uuid.uuid4().hex
else:
    raise RuntimeError("...")
```

should work for that case

hristog commented:
Thanks for looking into this PR, @jrbourbeau!

I've scaled down the functionality, associated with the newly-proposed config, as suggested, and updated its value to `bool`. The accompanying unit-tests have been updated accordingly, too.

One tiny concern I've got about the naming (`allow-random`) is that - as mentioned in [this comment](https://github.com/dask/dask/issues/6555#issuecomment-801116347) of mine - there are other sources of non-determinism that aren't going to be impacted by the proposed logic.

I suppose this could always be added, given actual demand, if GitHub issues are submitted requesting those to be supported at any point later on.

When you get a chance, please, let me know what your thoughts on regarding the best possible way to document the newly-introduced config.

hristog commented:
Hi @jrbourbeau - I'm writing to check if you've had a chance to look into the further changes that I pushed in https://github.com/dask/dask/pull/7413/commits/41b135b4d085fc26a7cfe6ce7475c70e649ef89e, and whether you'd like any additional refinements to be done on top of that.
Thanks!

jrbourbeau commented on dask/base.py:
```diff
@@ -839,7 +839,17 @@ def normalize_object(o):
+    if config.get("tokenize.allow-random", True):
+        return uuid.uuid4().hex
+
+    raise RuntimeError(
+        f"Object {str(o)} doesn't have a `__dask_tokenize__` method and "
+        "cannot be normalized deterministically."
+    )
```
Just a small suggestion on the error message here. In particular, it would be nice to point users to a relevant section in the docs. 

```
Object {str(o)} cannot be deterministically hashed. Please see
https://docs.dask.org/en/latest/custom-collections.html#implementing-deterministic-hashing
for more information
```

hristog replied:
Yep, I'll update it when I get a chance.

jrbourbeau commented on dask/tests/test_base.py:
```diff
@@ -224,6 +226,49 @@ def test_normalize_base():
         assert normalize_token(i) is i
 
 
+def test_normalize_object():
```
This test can be simplified a bit. In particular, we can probably just test that an informative error is raised when `tokenize.allow-random` is `False` and we attempt to `tokenize` an object which can't be deterministically hashed. See this comment https://github.com/dask/dask/pull/7413/files#r596499598 for an outline of a simplified test.

hristog replied:
Hmm.. my concerns here are that, without tests proving that true negatives are handled as such, we may not have sufficient confidence that the newly introduced functionality doesn't break behavior which it shouldn't affect. And, somehow, I don't feel comfortable to rely on other tests, external to this one, to do this for us.

Of course, if you strongly believe we could make do without the extra set of tests, I'll have to remove them (no issues about that). Just wanted to share my concerns.

hristog replied:
> This is a great point -- if you want to add a check for `tokenize.allow-random` in those locations too, that would be welcome

Hi @jrbourbeau - yes, I've opted to implement configurable behavior for those as well. The commits have already been added to this PR.

hristog replied:
> This test can be simplified a bit.

I've made an attempt at simplifying the tests. Hopefully, looking better from your point of view now. Let me know, please, if you'd like anything else changed.

jrbourbeau reviewed (commented):
Thanks @hristog -- apologies for the delayed response. 

> One tiny concern I've got about the naming (allow-random) is that - as mentioned in this comment of mine - there are other sources of non-determinism that aren't going to be impacted by the proposed logic.

This is a great point -- if you want to add a check for `tokenize.allow-random` in those locations too, that would be welcome

hristog commented:
Hi @jrbourbeau, I'm writing to check if you've had a chance to have a look at the most recent updates.
As discussed, I've scaled down the tests a bit further - the summary of my updates can be found in the most recent two comments, from [this thread](https://github.com/dask/dask/pull/7413#discussion_r609922956).

Please, let me know if the tests should be further stripped down, despite the discussed concerns.

ncclementi commented:
@jrbourbeau checking in here, it looks like @hristog updated the code according to your requests and CI is green. Do you think this is ready to go, or are there any other issues that should be addressed?

jsignell reviewed (commented):
This is looking good, you just need a change to the config file to support this change.

jcrist commented:
I've renamed the config field to `tokenize.ensure-deterministic` to better reflect the meaning of the parameter. I've also added it to the `dask.yaml` config file and the `dask-schema.yaml` schema description. I believe this should be good to go once tests pass.

hristog commented:
> I've renamed the config field to `tokenize.ensure-deterministic` to better reflect the meaning of the parameter. I've also added it to the `dask.yaml` config file and the `dask-schema.yaml` schema description. I believe this should be good to go once tests pass.

Thanks, @jcrist!

jrbourbeau commented:
Thanks for the updates @jcrist! I pushed a small commit to avoid `mock`ing in the tests added here. Will merge after CI finishes

jcrist commented:
Test failure is unrelated - merging. Thanks @hristog!

hristog commented:
> Thanks for the updates @jcrist! I pushed a small commit to avoid `mock`ing in the tests added here. Will merge after CI finishes

Thanks, @jrbourbeau! Thanks to your commit, I can much better understand what you meant in your [previous](https://github.com/dask/dask/pull/7413#discussion_r609922956) comments on this PR.
